Solve Gaussian Mixture Model sample/kmeans++-based initialisation merge conflicts #11101

alceballosa opened the pull request:
#### Reference Issues/PRs
Fixes conflicts in #11101. 

Resolves #10850 and #14398.

 The PR was stale for about 3 months so we introduced changes to account for other PRs involving the mixture model such as #17937 and #20030.

#### What does this implement/fix? Explain your changes.
PR #11101 was not merged due to a few conflicts and the fact that the kmeans_plusplus module wasn't public at the moment. We resolved the conflicts, so we modified the new functions so they referenced the public module introduced in #17937.

One key component of Mr. Gordon Walsh's implementation of GMM with sample and kmeans++-based initialization was to use a 0-iterations initialization. However, changes introduced in #20030 caused the 0-iterations instances to attempt to reach a line that should only be accessed with 1 or more iterations. For this reason, we introduced a conditional that avoids trying to do so when the max_iter parameter is 0.

Furthermore, there is a warning issued whenever the GMM was initialized with 0 iterations, since there is no convergence at that point. However, choosing to use max_iter=0 will usually be intentional, so we believe the user shouldn't be warned about non-convergence in such cases.

Finally, we modified the max_iter < 0 warning to reflect the fact that only negative numbers are not to be used, since previously the message would reference numbers smaller than 1 even if 0 was ok to use as the max_iter parameter.

#### Any other comments?

#DataUmbrella sprint
This PR was developed by @ariosramirez and myself. 
cc: @amueller @g-walsh

g-walsh committed:
Add data samples as initialisation. test to see if it is useful

g-walsh committed:
move the WIP test script

g-walsh committed:
Adjustment for PEP

g-walsh committed:
Add kmeans++ initialisation from kmeans module.

g-walsh committed:
Track indices in kmeans++ and edit kmeans and gmm implementation.

g-walsh committed:
remove test file

g-walsh committed:
replace original .gitignore

g-walsh committed:
Update for Flake8

g-walsh committed:
Add test for new initialisations in gmm. Check for consistent results.

g-walsh committed:
Update docstrings for base classes to include new initialisations.

g-walsh committed:
PEP8 changes and whitespace error in one docstring.

g-walsh committed:
Remove 'for' from description as it was highlighted in documentation.

g-walsh committed:
Update Documentation to include choice of initialization for gmm

g-walsh committed:
Change marker type for old matplotlib

g-walsh committed:
Fix inevitable PEP8 issues.

g-walsh committed:
Fix inevitable PEP8 issues.

g-walsh committed:
Remove camel case function names

g-walsh committed:
Add relative time taken to initialize to example and update docs

g-walsh committed:
Add relative time taken to initialize to example and update docs

g-walsh committed:
Update for ci errors. PEP and variable names

g-walsh committed:
Add iteration steps to plot.

g-walsh committed:
remove irrelevant comments and fix other review issues

g-walsh committed:
Max line length fix

g-walsh committed:
Add missing import

g-walsh committed:
Include iterations in documentation plot

g-walsh committed:
Alterations from review. 
- Documentation fixes
- docstring addition
- remove assert_true

g-walsh committed:
Allow gmm to take n_iter=0 for testing of initializations. Change plot to include this.

Now no longer need to import private _kmeans

g-walsh committed:
Clear up plot_gmm_init to remove confusing gen_gmm.

Rename a function and add comments.

g-walsh committed:
Resolve upstream conflict

g-walsh committed:
Merge remote-tracking branch 'upstream/master' into gmm_initialisation_improvement

g-walsh committed:
Add contibution to whats_new

g-walsh committed:
Typo

g-walsh committed:
Keep doc descriptions to one line.

g-walsh committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into gmm_initialisation_improvement

g-walsh committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into gmm_initialisation_improvement

g-walsh committed:
update tests

g-walsh committed:
Fix linting errors and renamed import

g-walsh committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into gmm_initialisation_improvement

g-walsh committed:
Add test for max_iter=0

alceballosa committed:
Solved merge conflicts for g-walsh's enhancements for the Gaussian Mixture Model PR#11101, focused on allowing sample-based and k-means++ based GMM initialization.. Also made the enhancements compatible with PRs #17937 and #20030.

alceballosa committed:
Merge remote-tracking branch 'upstream/main' into gmm_initialisation_improvement

alceballosa committed:
Fixed flake8 formatting problems

ogrisel commented on doc/whats_new/v0.24.rst:
```diff
@@ -82,6 +82,15 @@ Changelog
+  :class:`mixture.BayesianGaussianMixture` can now be initialized using
+  k-means++ and random data points. :issue:`11101` by
+  :user:`Gordon Walsh <g-walsh>`, conflicts solved by
+  :user:`Alberto Ceballos<alceballosa>` and Andres Rios<ariosramirez>.
```
This should be moved to `doc/whats_new/v1.0.rst` instead.

ogrisel commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1261,3 +1261,51 @@ def test_gaussian_mixture_setting_best_params():
+                              init_params=init, tol=1e-06, max_iter=1000)
+        gmm.fit(X)
+        assert_almost_equal(default_means, np.sort(gmm.means_.flatten()),
+                            decimal=1)
```
I don't find the `flatten` + `sort` combo convincing. I would rather compute the value of the value of the loglikelihood of the training set and compare that (`gmm.score(X)`).

Or alternatively, combine the pairwise cosine similarities of the means and check that the similarity matrix has one and only one entry with with a value larger than 0.99 in each column and the other values are below 0.01.

Side note: nowadays, we tend to use `assert_allclost` with the `atol` and `rtol` parameters to control for the precision is a more maintainable way (instead of the `decimal` parameter of `assert_almost_equal`).

alceballosa replied:
Thank you @ogrisel , I had a busy week but I will look into all of your suggestions ASAP :)

alceballosa replied:
Hi @ogrisel, I am attempting to implement the pairwise cosine similarity solution. The first condition works well, however, I don't seem to be getting values below 0.01 for the rest of the entries.

Here's a snippet of the proposed solution:

```
pw_sim = cosine_similarity(default_means, gmm.means_) #shape of both means: n_components, n_features
assert np.all(np.sum(pw_sim > 0.99, axis=0) == 1) and np.all(np.sum(pw_sim <= 0.01, axis=0) == pw_sim.shape[0]-1)

```
The test output indicates that most of the time the non >0.99 values have values greater than 0.01. Does that make sense to you?

ogrisel reviewed (commented):
Thanks for the follow-up. I am not convinced by the test though. See below.

See also the following suggestion to fix the changelog problem reported by the CI.

Maybe @jeremiedbb would like to have a look at this PR?

reshamas commented:
#DataUmbrella LATAM sprint

alceballosa committed:
Fixed what's new

alceballosa committed:
Fix mixture merge conflict

alceballosa committed:
Linting with black

alceballosa committed:
Fix linting and repeated section in kmeansplusplus docstring

alceballosa committed:
Black linting with correct version

alceballosa committed:
Additional black linting

alceballosa committed:
Fix introduced kmeans_plusplus index dtype error

alceballosa committed:
Fix changelog

alceballosa committed:
Reintroduced max_iter=0 verification

jjerphan commented on sklearn/mixture/_base.py:
```diff
@@ -252,28 +271,35 @@ def fit_predict(self, X, y=None):
+                    best_n_iter = n_iter
 
-        if not self.converged_:
+        # Should only warn about convergence if max_iter > 0, otherwise
+        # the user is assumed to have used 0-iters initialization
+        # to get the initial means.
+        if not self.converged_ and self.max_iter > 0:
```
Thanks for giving the motive. :+1:

jjerphan commented on sklearn/mixture/_base.py:
```diff
@@ -150,6 +152,23 @@ def _initialize_parameters(self, X, random_state):
             resp /= resp.sum(axis=1)[:, np.newaxis]
+        elif self.init_params == "rand_data":
+            resp = np.zeros((n_samples, self.n_components))
+            points = random_state.choice(
+                range(n_samples), self.n_components, replace=False
+            )
+            for n, i in enumerate(points):
+                resp[i, n] = 1
```
For consistency with the next branch of `k-means++`.
```suggestion
            indices = random_state.choice(
                range(n_samples), self.n_components, replace=False
            )
            for n, i in enumerate(indices):
                resp[i, n] = 1
```

jjerphan commented on sklearn/mixture/_bayesian_mixture.py:
```diff
@@ -118,13 +118,15 @@ class BayesianGaussianMixture(BaseMixture):
         The method used to initialize the weights, the means and the
         covariances.
-        Must be one of::
+        String must be one of::
```
```suggestion
        String must be one of:
```

jjerphan commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1258,72 @@ def test_gaussian_mixture_setting_best_params():
+
+def test_init_param():
+    # Check that the init_param options produces valid output.
+    # Compare all to default - kmeans.
```
```suggestion
    # Compare all to default (kmeans).
```

jjerphan commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1258,72 @@ def test_gaussian_mixture_setting_best_params():
+        gmm.fit(X)
+        default_means_init[init] = gmm.means_
+
+    for i_mean in default_means_init:
+        for j_mean in default_means_init:
+            if i_mean == j_mean:
+                pass
+            else:
+                assert np.any(
+                    np.not_equal(default_means_init[i_mean], default_means_init[j_mean])
+                )
```
Importing `itertools` at the top of the file, this can be simplified to:
```suggestion
    for i_mean, j_mean in itertools.combinations(default_means_init.values(), r=2):
        assert np.any(np.not_equal(i_mean, j_mean))
```

jjerphan reviewed (changes requested):
Thanks @alceballosa @ariosramirez for the follow-up and @g-walsh for the original PR.

Here are some comments and suggestions.

reshamas commented:
Sent an email reminder to @alceballosa and @ariosramirez in case GitHub notifications get buried.

alceballosa commented:
Thanks @jjerphan for the amazing feedback! I will review this during the weekend :)

alceballosa committed:
Update doc/modules/mixture.rst

More detailed documentation with more consistent itemization.  Minor grammar and other fixes for readability.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update examples/mixture/plot_gmm_init.py

Minor formatting changes, included Jake Vanderplas' GitHub user.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update doc/whats_new/v1.0.rst

Fixed username format for @ariosramirez

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update sklearn/mixture/_base.py

Keeping variable names consistent with the next branch of `k-means++`.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update sklearn/mixture/tests/test_gaussian_mixture.py

Minor symbol usage fix.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update sklearn/mixture/_bayesian_mixture.py

Changes for consistency in documentation.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update examples/mixture/plot_gmm_init.py

Change to inline of seed value used in only one instance.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update examples/mixture/plot_gmm_init.py

More adequate font size for non-full-screen visualization scenarios.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update examples/mixture/plot_gmm_init.py

Applied all caps for abbreviation (GMM).

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Update sklearn/mixture/tests/test_gaussian_mixture.py

Simplifying combinations using itertools.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into gmm_initialisation_improvement

alceballosa commented on examples/mixture/plot_gmm_init.py:
```diff
@@ -0,0 +1,106 @@
+
+
+# Author Gordon Walsh (github: g-walsh) gordon.p.walsh@gmail.com
+# data generation code from https://jakevdp.github.io/
```
We didn't, we only applied your suggestions. Should we mark ourselves in other files we did modify?

Thanks!

jjerphan replied:
I don't really know what the rules regarding adding authors to file headers are, so we better wait for confirmation from other maintainers.

alceballosa committed:
Added itertools import and removed seed declaration for consistency with previous commits

alceballosa commented on doc/whats_new/v0.21.rst:
```diff
@@ -81,7 +81,7 @@ Changelog
 :mod:`sklearn.inspection`
 .........................
 
```
Hi @jjerphan , what do you mean by 'patch' in this context? I resolved/committed all other relevant suggestions but I'm not sure about how to proceed with this one and the 'Ditto' suggestion. Thanks in advance!

jjerphan replied:
Hi @alceballosa, what I meant is that a few changes in this file are not related to your changes. You should have a whatsnew entry, but the others should not be changed, even if they aren't properly formatted.

Also, blank lines should not be added in files which are not semantically changed.

Regarding "Ditto", I wanted to indicate that this remark also apply there.

Let me know if something is still unclear. 🙂

alceballosa replied:
Hi @jjerphan, thanks a lot, I understand now. I see that I deleted a trailing whitespace to line 84 in the v0.21.rst file. Also, I see there is a blank line corresponding to one of my commits too in v0.24.rst.

What git command is the best for this situation? I guess I should go back to the commits prior to mine in those two files. I've looked into this tutorial https://docs.gitlab.com/ee/topics/git/numerous_undo_possibilities_in_git/ but I am not sure if that's the correct approach.

jjerphan replied:
That are several possibilities to modify previous commits (such as [using interactive rebase and fixup commits](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html)).

Here (in PRs which have been reviewed generally), we want to keep the published history; and as those patches have been introduced in various commit, the best and easiest is to just manually revert the changes an commit them all together.

Let me know if you need further guidance.

alceballosa replied:
Thanks @reshamas and @jjerphan. I just pushed the whitespace I deleted back in for the file v0.21.rst, line 84. Could you please check if it looks correct now?

As for file v0.24.rst, I am not sure about what to do there. There is a blank line which appears as if I had put it there but checking previous versions it seems it was already like that. I must have deleted and re-entered the line by mistake but it seems to me that I should just leave it like it is right now. Any thoughts?

Thank you once again :)

jjerphan replied:
`doc/whats_new/v0.24.rst` is unchanged now, and you should be able to revert the last spurious changes to `main`'s using this command:  
```sh
git checkout main -- doc/whats_new/v0.21.rst
```

Once it's done, it's mergeable to me. :slightly_smiling_face:

alceballosa replied:
Hi! I just did the checkout and pushed. Let me know if you can merge now.

alceballosa commented:
@jjerphan @reshamas Went over most suggestions today and commited changes. Pending some clarifications from @jjerphan I think it's all good. However, I'm getting a CircleCI test failed with the following message:

`ssh: connect to host github.com port 22: Connection timed out
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: pathspec 'pr/20408/merge' did not match any file(s) known to git
Could not fetch merge commit.
There may be conflicts in merging PR #20408 with main.`

Any ideas as to how I could fix this? I pulled from upstream but no merge conflicts were apparent.

Thank you once again,

Alberto

jjerphan commented:
This Circle CI error looks unrelated to your changes (it's a ssh error, which might be due to Circle CI infrastructure).

When the CI will be triggered again, this will get resolved.

jjerphan commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1259,66 @@ def test_gaussian_mixture_setting_best_params():
+    )
+    gmm.fit(X)
+
+    assert_array_equal(gmm.means_, means_init)
```
`assert_allclose` is more adapted when comparing floating.
```suggestion
    assert_allclose(gmm.means_, means_init)
```

alceballosa committed:
Update sklearn/mixture/tests/test_gaussian_mixture.py

Using more adequate function for floating point number equality comparisons.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

alceballosa committed:
fixup! Solved merge conflicts for g-walsh's enhancements for the Gaussian Mixture Model PR#11101, focused on allowing sample-based and k-means++ based GMM initialization.. Also made the enhancements compatible with PRs #17937 and #20030.

alceballosa committed:
Merge branch 'gmm_initialisation_improvement' of https://github.com/alceballosa/scikit-learn into gmm_initialisation_improvement

alceballosa committed:
Undid changes to doc/whats_new/v0.21.rst

jjerphan commented on doc/modules/mixture.rst:
```diff
@@ -135,6 +135,42 @@ parameters to maximize the likelihood of the data given those
+There is a choice of four initialization methods (as well as inputting user defined
+initial means) to generate the initial centers for the model components: 
+
+k-means (default)
+  This applies a traditional k-means clustering algorithm. This can be computationally expensive compared to other initialization methods.
```
```suggestion
k-means (default)
  This applies a traditional k-means clustering algorithm.
  This can be computationally expensive compared to other initialization methods.
```

alceballosa replied:
Thanks to you for your valuable feedback, this has been quite the learning opportunity for us :)

jjerphan reviewed (approved):
LGTM, thank you for pursuing this contribution @alceballosa and @ariosramirez!

One last suggestion for readability.

alceballosa committed:
Update doc/modules/mixture.rst

Better readability for the default k-means-based initialization in the documentation.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan commented:
Note that this PR might also resolve #10850 and #14398 and close #15744.

alceballosa commented:
Thanks @jjerphan. I will review them and update the original message accordingly.

reshamas commented:
Hi @alceballosa & @jjerphan 
I am not sure if we are waiting for contributor to address changes or for a core contributor to review. How to tell?

jjerphan commented:
Hi @reshamas,

This PR needs two core-devs (hence two others reviewers) to approve it before its branch can be merged (most of them are currently busy for releasing 1.0). Other reviewers might ask contributors to perform changes.

I just have added the "Waiting for Reviewer" label to help this PR get attention.

glemaitre commented:
I will do a review next week. @jeremiedbb might want to have a look as well since kmean++ is in the loop.

glemaitre reviewed (commented):
This is the first round of review. I will now look at the new example. It would be great to modify it to have a notebook-style example. It might then be better to avoid creating functions (it is a bit less intimidating for beginners) and add more text between cells to explain what do we do.

alceballosa committed:
Update sklearn/mixture/_base.py

Removed explicit for loop per @glemaitre.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

alceballosa committed:
Update sklearn/mixture/_base.py

Removed explicit for loop per @glemaitre.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

alceballosa committed:
Update sklearn/mixture/_bayesian_mixture.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

cmarmo commented:
Hi @alceballosa and @@ariosramirez, thanks a lot for your follow-up.
Right now there is a lint error in the pull request that does not allow to run the rest of the checks.
The error is
```
sklearn/mixture/_base.py:169:64: E999 SyntaxError: invalid syntax
            resp(np.arange(n_samples), indices.astype(np.int64)] = 1
                                                               ^
```
Do you mind fixing it?
Then @jeremiedbb might want to help with a second review?
Thanks!

ariosramirez committed:
Merge branch 'gmm_initialisation_improvement' of https://github.com/alceballosa/scikit-learn into gmm_initialisation_improvement

ariosramirez committed:
Solved SyntaxError caused by a typo into sklearn/mixture/_base.py

ariosramirez commented:
Hello everyone, thanks for you comments. 

I will be solving the latest issues. To be honest, I am learning how Sklearn works well, hence the delay.

I was testing the PR and the results are not good, you can try running `examples/mixture/plot_gmm_init.py`.
<img width="584" alt="Screen Shot 2021-12-03 at 09 47 37" src="https://user-images.githubusercontent.com/51466618/144605037-c4631fc0-f87d-4e29-9f8a-4575924a90d0.png">


**My focus will be on solve this. Thanks again for your patience.**

reshamas commented:
@ariosramirez 

In case you have questions on the PR. Can you make it to the next office hours for scikit-learn? Info is below.

---

Hi all,

Some of us will be online on the scikit-learn discord Monday December 6 at
10:00 PT / 13:00 ET / 18:00 UTC / 19:00 CET for about an hour or so.

First time and occasional contributors are welcome to join us to
discord using this invitation link:
https://discord.gg/YyYRXMju https://discord.gg/YyYRXMju

The focus of these office hour sessions is to answer questions about
contributing to scikit-learn. We can also split into break out
audio/text channels and do pair programming or live reviewing of
forgotten pull requests with screen sharing.

We can also try to assist you into crafting minimal reproduction cases
for bug reports to get a higher likelihood of resolution (e.g.
https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Please note, our Code of Conduct applies:
https://github.com/scikit-learn/scikit-learn/blob/main/CODE_OF_CONDUCT.md https://github.com/scikit-learn/scikit-learn/blob/main/CODE_OF_CONDUCT.md

ariosramirez committed:
solved bad rand_data and k-means++ initialisation

ariosramirez committed:
added rand_data and k-means++ description in GaussianMixture docstring

ariosramirez commented on sklearn/mixture/_bayesian_mixture.py:
```diff
@@ -118,13 +118,15 @@ class BayesianGaussianMixture(BaseMixture):
 
             'kmeans' : responsibilities are initialized using kmeans.
             'random' : responsibilities are initialized randomly.
+            'rand_data' : Initial means are randomly selected data points.
```
Hi @glemaitre I think `random_from_data` looks good to be explicit.

ariosramirez committed:
replace init_params from  rand_data to random_from_data

ariosramirez commented on sklearn/mixture/_base.py:
```diff
@@ -96,10 +98,10 @@ def _check_initial_parameters(self, X):
             )
 
-        if self.max_iter < 1:
+        if self.max_iter < 0:
```
Thanks @glemaitre, i'm going to implement using `sklearn/feature_extraction/text.py` as an example


```Python
def _validate_params(self):
      """Validation of min_df, max_df and max_features"""
      super()._validate_params()
    
      if self.max_features is not None:
          check_scalar(self.max_features, "max_features", numbers.Integral, min_val=0)
    
      if isinstance(self.min_df, numbers.Integral):
          check_scalar(self.min_df, "min_df", numbers.Integral, min_val=0)
      else:
          check_scalar(self.min_df, "min_df", numbers.Real, min_val=0.0, max_val=1.0)
    
      if isinstance(self.max_df, numbers.Integral):
          check_scalar(self.max_df, "max_df", numbers.Integral, min_val=0)
      else:
          check_scalar(self.max_df, "max_df", numbers.Real, min_val=0.0, max_val=1.0)
```

ariosramirez committed:
replace custom errors to sklearn.utils.check_scalar

ariosramirez committed:
delete messages validation because are tested in sklearn/utils/tests/test_validation.py

ariosramirez committed:
made the graph of example of plot_gmm_init bigger

ariosramirez committed:
changed the for loop to direct array replace

ariosramirez committed:
Merge branch 'main' into gmm_initialisation_improvement

ariosramirez committed:
parameterized init_params into test_init_param

ariosramirez committed:
changed assert_almost_equal param 'decimal' from 1 to 10

ariosramirez committed:
applied changes from black

ariosramirez committed:
solved circleci linting error

ariosramirez committed:
Merge branch 'scikit-learn:main' into gmm_initialisation_improvement

ariosramirez committed:
Merge branch 'gmm_initialisation_improvement' of https://github.com/alceballosa/scikit-learn into gmm_initialisation_improvement

ariosramirez committed:
applied black to test_gaussian_mixture.py

ariosramirez committed:
Merge branch 'gmm_initialisation_improvement' of https://github.com/ariosramirez/scikit-learn into gmm_initialisation_improvement

ariosramirez committed:
split test_init_param for each test case

ariosramirez committed:
Merge branch 'gmm_initialisation_improvement' of https://github.com/ariosramirez/scikit-learn into gmm_initialisation_improvement

ariosramirez commented:
Hello Team, the changes were applied (cc: @glemaitre) and the result is as expected (#11101)

![image](https://user-images.githubusercontent.com/51466618/146116606-56cb58ca-1d05-4118-880e-2692753e0644.png)


If anyone could give us feedback @ogrisel @thomasjpfan @jeremiedbb

Thanks!

cmarmo commented on sklearn/mixture/_gaussian_mixture.py:
```diff
@@ -492,13 +492,14 @@ class GaussianMixture(BaseMixture):
         precisions.
-        Must be one of::
-
+        String must be one of::
```
```suggestion
        String must be one of:
```
This will fix your sphinx warning and make all your checks green.

ariosramirez committed:
Update sklearn/mixture/_gaussian_mixture.py docstring

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

ariosramirez committed:
Update _gaussian_mixture.py docstring

reshamas commented:
@thomasjpfan @glemaitre @jeremiedbb
This PR is from the June 2021 Data Umbrella LATAM sprint.
Is there still work that needs to be done on this?

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -150,6 +139,21 @@ def _initialize_parameters(self, X, random_state):
+        elif self.init_params == "random_from_data":
+            resp = np.zeros((n_samples, self.n_components))
+            indices = random_state.choice(
+                range(n_samples), self.n_components, replace=False
```
range is not necessary, an integer is interpreted as such
```suggestion
                n_samples, size=self.n_components, replace=False
```

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -150,6 +139,21 @@ def _initialize_parameters(self, X, random_state):
+            _, indices = kmeans_plusplus(
+                X,
+                self.n_components,
+                x_squared_norms=row_norms(X, squared=True),
```
You can leave it unset, it will be computed automatically
```suggestion
```

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -150,6 +139,21 @@ def _initialize_parameters(self, X, random_state):
+                x_squared_norms=row_norms(X, squared=True),
+                random_state=random_state,
+            )
+            resp[indices.astype(int), np.arange(self.n_components)] = 1
```
Indices should already be integers. If it's not the case it's probably a bug
```suggestion
            resp[indices, np.arange(self.n_components)] = 1
```

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -252,28 +256,35 @@ def fit_predict(self, X, y=None):
+                    best_n_iter = n_iter
 
-        if not self.converged_:
+        # Should only warn about convergence if max_iter > 0, otherwise
+        # the user is assumed to have used 0-iters initialization
+        # to get the initial means.
+        if not self.converged_ and self.max_iter > 0:
```
you could just put this block into the `else` block so that it can only be triggered if max_iter > 0

alceballosa replied:
Hey, thanks for reviewing our PR!

The benefit of doing this doesn't seem as clear to me. I understand that doing a for-else will result in the same behavior but it seems to me that, in its current state, the warning will only get triggered if max_iter > 0 anyway. Could you clarify? I might have misunderstood your suggestion, too.

Kindly,

Alberto

jeremiedbb replied:
sorry I missed 1 level of indentation. It's fine as is indeed.

jeremiedbb commented on sklearn/mixture/_gaussian_mixture.py:
```diff
@@ -492,13 +492,14 @@ class GaussianMixture(BaseMixture):
         The number of initializations to perform. The best results are kept.
 
-    init_params : {'kmeans', 'random'}, default='kmeans'
+    init_params : str, default='kmeans'
```
Please write the set of accepted inits

jeremiedbb commented on sklearn/mixture/_bayesian_mixture.py:
```diff
@@ -118,13 +118,15 @@ class BayesianGaussianMixture(BaseMixture):
 
-    init_params : {'kmeans', 'random'}, default='kmeans'
+    init_params : {'kmeans', 'kmeans++', 'random', 'random_from_data'}, default='kmeans'
         The method used to initialize the weights, the means and the
         covariances.
-        Must be one of::
+        String must be one of:
 
             'kmeans' : responsibilities are initialized using kmeans.
             'random' : responsibilities are initialized randomly.
+            'random_from_data' : Initial means are randomly selected data points.
+            'k-means++' : Use the k-means++ method to initialize.
```
Please add a `versionchanged:: 1.1` field saying that this params now accepts 'random_from_data' and 'k-means++'.

jeremiedbb commented on sklearn/mixture/_gaussian_mixture.py:
```diff
@@ -492,13 +492,14 @@ class GaussianMixture(BaseMixture):
         The method used to initialize the weights, the means and the
         precisions.
-        Must be one of::
-
-            'kmeans' : responsibilities are initialized using kmeans.
-            'random' : responsibilities are initialized randomly.
+        String must be one of:
+        - 'kmeans' : responsibilities are initialized using kmeans.
+        - 'random' : responsibilities are initialized randomly.
+        - 'random_from_data' : Initial means are randomly selected data points.
+        - 'k-means++' : Use the k-means++ method to initialize.
```
Please add a versionchanged:: 1.1 field saying that this params now accepts 'random_from_data' and 'k-means++'.

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -127,56 +128,39 @@ def test_gaussian_mixture_attributes():
-        "Estimation requires at least one component"
-    )
-    with pytest.raises(ValueError, match=msg):
+
+    with pytest.raises(ValueError):
```
I'd rather keep checking the error message, here and below as well. It might have change after switching to use `check_scalar` in the parameter checks.

alceballosa replied:
@ariosramirez can you check this suggestion? Thanks!

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1242,71 @@ def test_gaussian_mixture_setting_best_params():
+        max_iter=1000,
+    )
+    gmm.fit(X)
+    assert_almost_equal(default_means, np.sort(gmm.means_.flatten()), decimal=10)
```
It's better to use `assert_allclose` which checks relative difference

alceballosa replied:
I used ```assert_allclose``` with ```rtol=1e-8```. Does that sound good or should I constrain it a bit more?

jeremiedbb replied:
Actually you can leave rtol to its default value, which is 1e-7. This is what we consider acceptable throughout the project.

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1242,71 @@ def test_gaussian_mixture_setting_best_params():
+        assert np.any(np.not_equal(i_mean, j_mean))
+
+
+def test_check_that_max_iter_with_zero_returns_initialisation_as_expected():
```
```suggestion
def test_max_iter_zero():
```

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1258,3 +1242,71 @@ def test_gaussian_mixture_setting_best_params():
+@pytest.mark.parametrize(
+    "init_params", ["random", "random_from_data", "k-means++", "kmeans"]
+)
+def test_init_param(init_params):
+    # Check that the init_param options produces valid output.
+    # Compare all to default (kmeans).
```
I think this test is brittle since afaik there's no guarantee that different inits will converge to the same solution (even up to a permutation). I'd be in favor of not adding such a test

alceballosa replied:
Should we delete this test outright or leave it for another PR? FWIW, I think you are right but I'd have to read a bit/ask around to check whether that assumption on GMM convergence under different initialization methods is valid.

jeremiedbb replied:
I checked with a wide range of random seeds at it fails from time to time which confirms that it's not a property we can assume.  So I definitely think that we should not include it in this PR

alceballosa replied:
Thanks! Just removed it.

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -252,28 +256,35 @@ def fit_predict(self, X, y=None):
-                prev_lower_bound = lower_bound
+            if self.max_iter == 0:
+                best_params = self._get_parameters()
+                best_n_iter = 1
```
shouldn't it be `best_n_iter = 0` ?

alceballosa replied:
You're right, thanks for pointing this out.

jeremiedbb reviewed (commented):
Thanks @alceballosa for working on this.

I wonder if we really want to keep the "random" init. It seems like a really bad init strategy. I'd be in favor of changing the behavior of "random" (with a deprecation cycle ?) to be the one defined in "random_from_data". I'd like to know what others think about that.

I also made a few comments

ariosramirez commented:
Thank you for your comments @jeremiedbb.

Just a question, what examples of the deprecation cycle can I follow?

jeremiedbb commented:
> Just a question, what examples of the deprecation cycle can I follow?

Here's the doc on how to implement that https://scikit-learn.org/stable/developers/contributing.html#maintaining-backwards-compatibility.

But thinking about that again, I think we should not do that in this PR. Let's keep both behaviors for now. It makes the PR easier to review hoping it'll get merged more quickly :) We can think about possible deprecation in a separate issue/PR.

jeremiedbb commented on doc/whats_new/v1.0.rst:
```diff
@@ -908,6 +908,13 @@ Changelog
 
+- |Feature| :class:`mixture.GaussianMixture` and
+  :class:`mixture.BayesianGaussianMixture` can now be initialized using
+  k-means++ and random data points. :issue:`11101` by
+  :user:`Gordon Walsh <g-walsh>`, conflicts solved in :pr:`20408` by
+  :user:`Alberto Ceballos<alceballosa>` and
+  :user:`Andres Rios<ariosramirez>`.
```
Also, the changelog entry should be in v1.1

alceballosa replied:
Fix changelog entry, remove unnecessary 11101 issue ref.

alceballosa committed:
Update sklearn/mixture/_base.py

Omit unnecessary use of the range function.

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

alceballosa committed:
Update sklearn/mixture/_base.py

Removing unnecessary computation of x_squared_norms.

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

alceballosa committed:
Update sklearn/mixture/tests/test_gaussian_mixture.py

Simplified zero iterations initialization test name.

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

alceballosa committed:
Merge gaussian mixture test conflict

alceballosa committed:
Update doc/whats_new/v1.0.rst

Fix changelog entry, remove unnecessary 11101 issue ref.

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

alceballosa committed:
Moved contrib entry from v1.0.rst to v1.1.rst

alceballosa committed:
Removed unnecessary cast to int

alceballosa committed:
Improved init_params documentation in bayesian and gaussian mixture

alceballosa committed:
Switched to assert_allclose for testing init params

alceballosa committed:
Set best_n_params correctly when max_iter equals 0

alceballosa committed:
Remove unnecessary row_norms import

alceballosa commented:
I'm wrapping up the suggestions made by @jeremiedbb. However, there seems to be a linting issue (check: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=39801&view=logs&j=32e2e1bb-a28f-5b18-6cfc-3f01273f5609&t=fc67071d-c3d4-58b8-d38e-cafc0d3c731a) related to black. 

I'm using black 21.6b0 as per the Contributing documentation and running the black command as ```black -t py39 <filename>.py```. As per the documentation, it should be ```black <filename.py>```, however, using Python 3.9.10 (installed by conda when setting python=3.9) seems to make black think that the target is 3.10:

```
Error: Invalid value for '-t' / '--target-version': 'py310' is not one of 'py27', 'py33', 'py34', 'py35', 'py36', 'py37', 'py38', 'py39'.
```

Should I downgrade to Python 3.9.0 or something? @cmarmo @jjerphan 

Thanks!

jeremiedbb commented:
We switched to ``black==22.1.0``. It's in the dev version of the contributing guide. We are trying to make the contributing guide from the stable version of the doc to point to the dev version.

alceballosa committed:
Reformat with black 22.1.0, changed rtol to standard value

jeremiedbb commented on sklearn/mixture/_bayesian_mixture.py:
```diff
@@ -118,13 +118,20 @@ class BayesianGaussianMixture(BaseMixture):
         lower bound value on the likelihood is kept.
 
-    init_params : {'kmeans', 'random'}, default='kmeans'
+    init_params : {'kmeans', 'k-means++', 'random', 'random_from_data'},
+    default='kmeans'
```
If it does not fit in one line, you can break the line
```suggestion
    init_params : {'kmeans', 'k-means++', 'random', 'random_from_data'}, \
            default='kmeans'
```

jeremiedbb commented on sklearn/mixture/_gaussian_mixture.py:
```diff
@@ -469,7 +469,8 @@ class GaussianMixture(BaseMixture):
         The number of mixture components.
 
-    covariance_type : {'full', 'tied', 'diag', 'spherical'}, default='full'
+    covariance_type : {'full', 'tied', 'diag', 'spherical'},
+    default='full'
```
```suggestion
    covariance_type : {'full', 'tied', 'diag', 'spherical'}, default='full'
```

jeremiedbb commented on sklearn/mixture/_gaussian_mixture.py:
```diff
@@ -492,13 +493,20 @@ class GaussianMixture(BaseMixture):
         The number of initializations to perform. The best results are kept.
 
-    init_params : {'kmeans', 'random'}, default='kmeans'
+    init_params : {'kmeans', 'k-means++', 'random', 'random_from_data'},
+    default='kmeans'
```
```suggestion
    init_params : {'kmeans', 'k-means++', 'random', 'random_from_data'}, \
            default='kmeans'
```

jeremiedbb reviewed (commented):
Here are some hints to fix the CI failures

alceballosa committed:
Init params test assumes GMM can converge to the same solution, which is not the case. Decided to remove this test.

alceballosa committed:
Fixed CI failures caused by docstrings

alceballosa commented:
Thanks once again @jeremiedbb. It seems that the only remaining issues are related to the ```pytest.raises``` messages but as I didn't do the commit removing the messages I'm not sure how to proceed. I suppose it will be ok just to put them back in as they were before Andres' commit, but I noticed you mention ```check_scalar``` and I can't find any reference to it.

jeremiedbb commented on sklearn/mixture/_base.py:
```diff
@@ -76,40 +78,26 @@ def _check_initial_parameters(self, X):
-                "Invalid value for 'n_init': %d Estimation requires at least one run"
-                % self.n_init
-            )
+        check_scalar(
+            self.n_init, name="n_init", target_type=numbers.Integral, min_val=1
+        )
```
This change will modifiy the error message because ``check_scalar`` has a different wording.
It needs to be reflected in this test (`test_gaussian_mixture_attributes`)
```py
    n_init_bad = 0
    gmm = GaussianMixture(n_init=n_init_bad)
    msg = (
        f"Invalid value for 'n_init': {n_init_bad} Estimation requires at least one run"
    )
    with pytest.raises(ValueError, match=msg):
        gmm.fit(X)
```

alceballosa replied:
Would this work for that specific case?

```python
    msg = (
        f"n_init == {n_init_bad}, must be >= 1."
    )
```
I tried to use some a more verbose message but it seems that ```check_scalar``` messages are expected to be in the form ```'n_init == 0, must be >= 1.'```

jeremiedbb replied:
Yes that's what is expected

alceballosa replied:
Thanks @jeremiedbb. I just added the remaining cases!

On another note, it seems that removing the init_params test reduced coverage and codecov is now not passing. Should we devise a new test for that purpose?

reshamas commented:
@alceballosa
This might be a helpful reference for the `check_scalar` function, with example PRs:
https://github.com/scikit-learn/scikit-learn/issues/21927

alceballosa committed:
Matched error messages from check_scalar

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1263,6 +1250,46 @@ def test_gaussian_mixture_setting_best_params():
+@pytest.mark.parametrize(
+    "init_params", ["random", "random_from_data", "k-means++", "kmeans"]
+)
+def test_check_non_unique_means_initialisation(init_params):
+    # Check that all initialisations provide unique starting means
```
I'm not sure what this test is supposed to check. Is it that different init_params lead to all different init means ? I don't think we can ensure that. It's unlikely, but it might happen for some random seed.

I think it would be better to check that general properties hold for each init_params.

alceballosa replied:
I think the purpose is rather to ensure that the means within each initialization are different, which makes a bit more sense.  Do you still think it's better to remove it? I could change the comment to be more explicit and add your suggested general properties test as a new test.

jeremiedbb replied:
(Sorry, I just noticed that you answered).

> I think the purpose is rather to ensure that the means within each initialization are different

That would be acceptable to test indeed. But that's not what the current code is doing.

>  I could change the comment to be more explicit and add your suggested general properties test as a new test.

Sure. I think it can be done in this same test.

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1263,6 +1250,46 @@ def test_gaussian_mixture_setting_best_params():
+        n_components=n_components, random_state=0, init_params=init_params, max_iter=1
+    )
+    gmm.fit(X)
+    default_means_init[init_params] = gmm.means_
+    for i_mean, j_mean in itertools.combinations(default_means_init.values(), r=2):
+        assert np.any(np.not_equal(i_mean, j_mean))
```
Following my comment above, we could instead check that
- the fitted means have the expected shape
- the fitted means are "within" the data, i.e. between xmin and xmax for each feature
- it managed to converge, i.e. ``converged_`` is True
```suggestion

assert gmm.means_.shape == (n_components, n_features)
assert np.all(X.min(axis=0) <= gmm.means_)
assert np.all(gmm.means_ <= X.max(axis=0))
assert gmm.converged_
```

jeremiedbb commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -1263,6 +1250,46 @@ def test_gaussian_mixture_setting_best_params():
+    X = rand_data.X["full"]
+    default_means_init = {}
+
+    gmm = GaussianMixture(
+        n_components=n_components, random_state=0, init_params=init_params, max_iter=1
+    )
```
```suggestion
    gmm = GaussianMixture(
        n_components=n_components, random_state=0, init_params=init_params
    )
```

jeremiedbb reviewed (commented):
I think one the test is still testing stuff we can't guarantee. I propose a few changes to improve the testing of the different inits

jeremiedbb commented:
'k-means++' is already the name of an init of KMeans. I think it's important to keep the same name for both.

>  I can see some people trying to change from 'kmeans' to the ++ version and getting an error by not putting the dash.

They'll still get an informative error message giving all the possibilities, so I think's fine.

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/alceballosa/20408

jeremiedbb committed:
solve conflicts, fix test, add test

jeremiedbb committed:
cln

jeremiedbb reviewed (commented):
LGTM. Thanks @alceballosa and @ariosramirez !

jeremiedbb committed:
cln what's new

jeremiedbb committed:
re cln what's new

jeremiedbb committed:
same

reshamas commented:
Congrats @alceballosa @ariosramirez for continuing work on this after the LATAM sprint!

Thanks to all the maintainers too for this!
@thomasjpfan We can update the sprint list too :)
cc:  @amueller
